iceoryx_posh/test/fuzztests/README.md

```diff
@@ -0,0 +1,49 @@
+
+The fuzz wrappers currently support three different interfaces: the unix domain socket communication, the message parsing method and the toml configuration file parser.  
+1. The unix domain socket can either be tested by starting RouDi and sending messages to RouDi via uds or by directly invoking the message parsing function within RouDi. The first solution has the advantage, that the messages from uds to RouDi take the same path as normal messages from applications would do. The drawback however is that RouDi needs to be started with all the overhead which would not be necessary for this test which therefore slows the test a little bit.
+
+2. By invoking the message parsing function directly, the method is independent from the underlying protocol such as uds. It should also be slightly faster since some functions are not invoked compared to uds fuzzing. However, a RouDi thread is also started with this approach because otherwise it wouldn't be possible to invoke the message process function within RouDi without directly modifying the code in RouDi. 
+
+3. As a third use-case, the toml configuration parser can be tested. An example of a toml file can be found here: \[path\_to\_iceoryx\]/iceoryx\_posh/etc/iceoryx/roudi\_config\_example.toml. If you chose this method, you also need to set -t or --toml-file <PATH\_TO\_FILE>. A file needs to be specified which can be used to write in the messages which will send to the toml configuration parser. 
+
+The interface you want to fuzz can be chosen with -f or --fuzzing-api + entering the argument: uds, com or toml. The argument uds is used for use-case 1, com for use-case 2 and toml for use-case 3. 
+
+The way how the fuzzer sends messages to the interface can also be specified via -m or --input-mode
+1. The messages can be sent via stdin (-m stdin)
+2. The messages can be sent via using the command line (-m cl) followed by -i or --command-line-input with the message following as argument or -c --command-line-file to specify a file where the message you want to send is located. By using -c or --command-line-file, the complete file is sent as one message and is not sent line by line. 
+
+Additionally, the log level can be chosen via -l or --log-level such that messages are printed for debug or not. The default is to switch logging off. 3 arguments are possible:
+1. off: to disable logging (standard case)
+2. fatal: log fatal messages
+3. debug: log debug messages
+
+As an example: afl can be started for uds with: afl-fuzz -i \[path\_to\_input\_example\] -o \[path\_where\_output\_of\_fuzzer\_is\_written\_to\] \[path\_to\_iceoryx\]/build/posh/test/posh\_fuzztests -f uds -m stdin
+ 
```

Comment:
There where some things i stumpled on when using the AFL:

1. The following error message came up in the first run:
```bash
[-] Whoops, your system uses on-demand CPU frequency scaling, adjusted
    between 781 and 4882 MHz. Unfortunately, the scaling algorithm in the
    kernel is imperfect and can miss the short-lived processes spawned by
    afl-fuzz. To keep things moving, run these commands as root:

    cd /sys/devices/system/cpu
    echo performance | tee cpu*/cpufreq/scaling_governor

    You can later go back to the original state by replacing 'performance' with
    'ondemand'. If you don't want to change the settings, set AFL_SKIP_CPUFREQ
    to make afl-fuzz skip this check - but expect some performance drop.

[-] PROGRAM ABORT : Suboptimal CPU scaling governor
         Location : check_cpu_governor(), afl-fuzz.c:7409
```

Setting the scaling_governor to performance solved this as described in the error text. A hint in the documentation would be nice.

2. The following error message came up in the second run:
```bash
[-] Oops, the program crashed with one of the test cases provided. There are
    several possible explanations:

    - The test case causes known crashes under normal working conditions. If
      so, please remove it. The fuzzer should be seeded with interesting
      inputs - but not ones that cause an outright crash.

    - The current memory limit (50.0 MB) is too low for this program, causing
      it to die due to OOM when parsing valid files. To fix this, try
      bumping it up with the -m setting in the command line. If in doubt,
      try something along the lines of:

      ( ulimit -Sv $[49 << 10]; /path/to/binary [...] <testcase )

      Tip: you can use http://jwilk.net/software/recidivm to quickly
      estimate the required amount of virtual memory for the binary. Also,
      if you are using ASAN, see /usr/local/share/doc/afl/notes_for_asan.txt.

    - Least likely, there is a horrible bug in the fuzzer. If other options
      fail, poke <lcamtuf@coredump.cx> for troubleshooting tips.

[-] PROGRAM ABORT : Test case 'id:000000,orig:fuzz_input.txt' results in a crash
         Location : perform_dry_run(), afl-fuzz.c:2898
```

Adding the command-line argument `-m 2000` solved this problem and the fuzzer runs successfully. Please add this option to your doc and give a short hint about this.

Reply:
@dkroenke Thank you very much for providing this information. However, the question for me is if we want to give specific information how to start afl-fuzz. The Fuzz-Wrappers are independent of the used fuzzer and maybe it's better to just provide a link to the fuzzer. For example, I haven't get the first error you described. Therefore, I'm not sure if it makes sense to give concrete examples how to solve afl problems.

Reply:
Strangely the fuzzer only starts when i attach `-m 2000` to the command, otherwise it is always throwing this error, tested that with fish and bash shell. The fuzzer stats don't indicate any error if I interpret them correctly, i attached them.

```
start_time        : 1620408641
last_update       : 1620408699
fuzzer_pid        : 1483995
cycles_done       : 1
execs_done        : 135721
execs_per_sec     : 1333.33
paths_total       : 43
paths_favored     : 6
paths_found       : 42
paths_imported    : 0
max_depth         : 4
cur_path          : 39
pending_favs      : 0
pending_total     : 32
variable_paths    : 0
stability         : 100.00%
bitmap_cvg        : 0.76%
unique_crashes    : 0
unique_hangs      : 0
last_path         : 1620408687
last_crash        : 0
last_hang         : 0
execs_since_crash : 135721
exec_timeout      : 20
afl_banner        : posh_fuzztests
afl_version       : 2.57b
target_mode       : default
command_line      : afl-fuzz -i ../../../iceoryx_posh/test/fuzztests/fuzz_input/ -o ./fuzz_output ./posh_fuzztests -f uds -m stdin -m 2000
slowest_exec_ms   : 1
peak_rss_mb       : 4
```